dev/com.ibm.ws.microprofile.health/src/com/ibm/ws/microprofile/health/internal/AppTrackerImpl.java

```diff
@@ -118,50 +96,141 @@ public void onStartup(Set<Class<?>> arg0, ServletContext ctx) throws ServletExce
         return appModules.get(appName);
     }
 
-    /*
-     * collect all app and module names and save it for later use
-     */
-    private AppModuleName setAppModuleNames(IServletContext isc) {
+    /** {@inheritDoc} */
+    @Override
+    @FFDCIgnore(UnableToAdaptException.class)
+    public void applicationStarting(ApplicationInfo appInfo) throws StateChangeException {
+        String appName = appInfo.getName();
+        if (tc.isDebugEnabled())
+            Tr.debug(tc, "applicationStarting() : appName = " + appInfo.getName());
+
+        Container appContainer = appInfo.getContainer();
+        if (appContainer == null) {
+            if (tc.isDebugEnabled()) {
+                Tr.debug(tc, "applicationStarting() : appContainer=null for " + appInfo);
+            }
+            return;
+        }
 
-        WebAppConfig webAppConfig = isc.getWebAppConfig();
-        if (webAppConfig.isSystemApp()) {
-            Tr.debug(tc, "Detected system app so won't track for health check; appName = ", webAppConfig.getApplicationName());
-            return null;
+        try {
+            NonPersistentCache cache = appContainer.adapt(NonPersistentCache.class);
+            ApplicationClassesContainerInfo acci = (ApplicationClassesContainerInfo) cache.getFromCache(ApplicationClassesContainerInfo.class);
```

Comment:
acci doesn't seem to be read anywhere else.

Reply:
We need to check if the ApplicationClassesContainerInfo is in the container cache, before getting the application info from the container. I do a null check for acci in 118, and return if the application classes can't be found. If the application classes are found, than our Application Manager is in control of the application. For OSGI apps, (such as .eba or .esa applications, the acci would be null.